### PR TITLE
Add NewType class to have strong typedefs for primitive types.

### PR DESCRIPTION
**Pull request by mgsergio**
@ygorshenin 
@milchakov

### CODE REVIEW

File: base/newtype.hpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include "std/type_traits.hpp"
+
+namespace my
+{
+namespace impl
+{
+template <typename From, typename To>
```
Comment (ygorshenin) [inline base/newtype.hpp]: TFrom, TTo

File: base/newtype.hpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+using IsConvertibleGuard = typename enable_if<is_convertible<From, To>::value>::type*;
```
Comment (ygorshenin) [inline base/newtype.hpp]: clang-format, please.

File: base/newtype.hpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+}  // namespace impl
+
+template <typename U, typename Tag>
```
Comment (ygorshenin) [inline base/newtype.hpp]: U -> TType
Tag -> TTag
:)

Comment (ygorshenin) [inline base/newtype.hpp]: Comment to the class, please.

File: base/newtype.hpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+  NewType & operator++()
+  {
+    ++m_value;
+    return *this;
+  }
+
+  NewType operator++(int)
+  {
+    auto copy = *this;
```
Comment (ygorshenin) [inline base/newtype.hpp]: auto const

File: base/newtype.hpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+  NewType & operator--()
+  {
+    --m_value;
+    return *this;
+  }
+
+  NewType operator--(int)
+  {
+    auto copy = *this;
```
Comment (ygorshenin) [inline base/newtype.hpp]: auto const

File: base/newtype.hpp (hunk 2 above)
Comment (milchakov) [inline base/newtype.hpp]: Думаю, что эта проверка должна быть не только на конвертацию, но и на то, что типы интегральные 
is_integral<From> && is_integral<To>.

File: base/newtype.hpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    return *this;
+  }
+
+  NewType & operator++()
+  {
+    ++m_value;
+    return *this;
+  }
+
+  NewType operator++(int)
```
Comment (milchakov) [inline base/newtype.hpp]: const NewType operator++(int)

File: base/newtype.hpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    return copy;
+  }
+
+  NewType & operator--()
+  {
+    --m_value;
+    return *this;
+  }
+
+  NewType operator--(int)
```
Comment (milchakov) [inline base/newtype.hpp]: const NewType operator--(int)

File: base/newtype.hpp (hunk 6 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 7 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 5 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 4 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 2 above)
Comment (milchakov) [inline base/newtype.hpp]: обсудили, можно не добавлять

**Comment by mgsergio** [comment]
PTAL

File: base/newtype.hpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  {
+    m_value &= v.m_value;
+    return *this;
+  }
+
+  // TODO(mgsergio): Is it meaningful for a newtype to have << operator ?
+  // NewType & operator<<=(NewType<V, VTag> const & v)
+  // NewType & operator>>=(NewType<V, VTag> const & v)
+
+  bool operator==(NewType const & o)
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+  {
+    return m_value == o.m_value;
+  }
+
+  bool operator!=(NewType const & o)
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+  {
+    return !(m_value == o.m_value);
+  }
+
+  bool operator<(NewType const & o)
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  {
+    return m_value < o.m_value;
+  }
+
+  bool operator>(NewType const & o)
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/base_tests/newtype_test.cpp [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+NEWTYPE(int, Int);
+
+string DebugPrint(Int const & i)
+{
+  stringstream sstr;
+  sstr << "Int(" << i.Get() << ')';
+  return sstr.str();
+}
+
+UNIT_TEST(NewTypeTypeChecks)
```
Comment (ygorshenin) [inline base/base_tests/newtype_test.cpp]: NewType_TypeChecks

File: base/base_tests/newtype_test.cpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+{
+  TEST((is_constructible<Int, int>::value), ());
+  TEST((is_constructible<Int, char>::value), ());
+  TEST(!(is_convertible<int, Int>::value), ());
+  TEST(!(is_convertible<Int, int>::value), ());
+}
+
+UNIT_TEST(NewTypeBase)
```
Comment (ygorshenin) [inline base/base_tests/newtype_test.cpp]: NewType_Base

File: base/base_tests/newtype_test.cpp [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+{
+  Int a{10};
+  TEST_EQUAL(a.Get(), 10, ());
+
+  a.Set(100);
+  TEST_EQUAL(a.Get(), 100, ());
+}
+
+UNIT_TEST(NewTypeOperations)
```
Comment (ygorshenin) [inline base/base_tests/newtype_test.cpp]: NewType_Operations

File: base/base_tests/newtype_test.cpp [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+  TEST_EQUAL(Int(10) / Int(2), Int(5), ());
+  TEST_EQUAL(Int(10) * Int(2), Int(20), ());
+  TEST_EQUAL(Int(10) % Int(3), Int(1), ());
+
+  TEST_EQUAL(Int(10) | Int(7), Int(10 | 7), ());
+  TEST_EQUAL(Int(10) & Int(7), Int(10 & 7), ());
+  TEST_EQUAL(Int(10) ^ Int(7), Int(10 ^ 7), ());
+}
+
+UNIT_TEST(NewTypeMemberOperations)
```
Comment (ygorshenin) [inline base/base_tests/newtype_test.cpp]: NewType_MemberOperations

File: base/newtype.hpp (hunk 2 above)
Comment (mgsergio) [inline base/newtype.hpp]: clang-format is done.

File: base/base_tests/newtype_test.cpp (hunk 12 above)
Comment (mgsergio) [inline base/base_tests/newtype_test.cpp]: Done.

File: base/base_tests/newtype_test.cpp (hunk 13 above)
Comment (mgsergio) [inline base/base_tests/newtype_test.cpp]: Done.

File: base/base_tests/newtype_test.cpp (hunk 14 above)
Comment (mgsergio) [inline base/base_tests/newtype_test.cpp]: Done.

File: base/newtype.hpp [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+/// Creates a typesafe alias to a given numeric Type.
+template <typename Type, typename Tag>
+class NewType
+{
+public:
+  template <typename V, impl::IsConvertibleGuard<V, Type> = nullptr>
+  explicit NewType(V const & v) : m_value(v)
+  {
+    static_assert(is_integral<Type>::value || is_floating_point<Type>::value,
```
Comment (ygorshenin) [inline base/newtype.hpp]: Move this just below the public.

File: base/base_tests/newtype_test.cpp (hunk 15 above)
Comment (mgsergio) [inline base/base_tests/newtype_test.cpp]: Done.

File: base/newtype.hpp [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+  {
+    m_value &= v.m_value;
+    return *this;
+  }
+
+  // TODO(mgsergio): Is it meaningful for a newtype to have << operator ?
+  // NewType & operator<<=(NewType<V, VTag> const & v)
+  // NewType & operator>>=(NewType<V, VTag> const & v)
+
+  bool operator==(NewType const & o) { return m_value == o.m_value; }
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+  bool operator!=(NewType const & o) { return !(m_value == o.m_value); }
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+  bool operator<(NewType const & o) { return m_value < o.m_value; }
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+  bool operator>(NewType const & o) { return m_value > o.m_value; }
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+  bool operator<=(NewType const & o) { return !(m_value > o.m_value); }
```
Comment (ygorshenin) [inline base/newtype.hpp]: const

File: base/newtype.hpp [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+  bool operator>=(NewType const & o) { return !(m_value < o.m_value); }
+  NewType operator+(NewType const & o) { return NewType(m_value + o.m_value); }
```
Comment (ygorshenin) [inline base/newtype.hpp]: All these methods must be const.

File: base/newtype.hpp (hunk 11 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 10 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 8 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 9 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

File: base/newtype.hpp (hunk 22 above)
Comment (mgsergio) [inline base/newtype.hpp]: Done.

**Comment by milchakov** [comment]
LGTM

**Comment by mgsergio** [comment]
PTAL

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**